### PR TITLE
adapter,sql: add ALTER SINK statement

### LINKED ISSUE #15310: `ALTER SINK ... SIZE` support

### PR DESCRIPTION
**Pull request by andrioni**
Add ALTER SINK statement with support for two changeable parameters (REMOTE, SIZE) and one unchangeable parameter (SNAPSHOT). Code follows the same structure used by ALTER SOURCE, and supports the same set of operations (`SET`, `RESET`).

Closes #15310.

### Motivation

  * This PR adds a known-desirable feature: #15310

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Support changing the size of an existing sink via `ALTER SINK` statement.

### CODE REVIEW

File: src/sql-parser/tests/testdata/ddl [hunk 1]
```diff
@@ -898,9 +898,30 @@ ALTER MATERIALIZED VIEW name SET (property = true)
...
 ALTER SINK name SET (property = true)
 ----
-error: Expected RENAME, found SET
+error: Expected one of REMOTE or SIZE or SNAPSHOT, found identifier "property"
 ALTER SINK name SET (property = true)
-                ^
+                     ^
+
+parse-statement
+ALTER SINK name SET (property = true)
```
Comment (bkirwi) [inline src/sql-parser/tests/testdata/ddl:906]: Looks like a duplicate!

File: src/sql/src/plan.rs [hunk 2]
```diff
@@ -548,6 +550,21 @@ pub struct AlterIndexResetOptionsPlan {
     pub options: HashSet<IndexOptionName>,
 }
 
+#[derive(Debug, Clone)]
+
+pub enum AlterSinkItem {
```
Comment (bkirwi) [inline src/sql/src/plan.rs:555]: Could share this with the `Source`; it's ~identical. Unless you think they're likely to evolve to be more different?

**Review by bkirwi — COMMENTED** [review]
A few minor comments but looking good overall!

File: src/adapter/src/catalog.rs [hunk 3]
```diff
@@ -3460,6 +3460,91 @@ impl<S: Append> Catalog<S> {
...
+                    use mz_sql::ast::Value;
+                    use mz_sql_parser::ast::CreateSinkOptionName::*;
+                    use AlterSinkItem::*;
+
+                    let entry = state.get_entry(&id);
+                    let name = entry.name().clone();
+                    let old_sink = match entry.item() {
+                        CatalogItem::Sink(sink) => sink.clone(),
+                        other => {
+                            coord_bail!("ALTER SINK entry was not a source: {}", other.typ())
```
[the remark this answers is not in the document]
Comment (andrioni) [reply src/adapter/src/catalog.rs]: Thanks!

File: src/adapter/src/catalog.rs [hunk 4]
```diff
@@ -3460,6 +3460,91 @@ impl<S: Append> Catalog<S> {
...
+                    let mut stmt = mz_sql::parse::parse(&old_sink.create_sql)
+                        .unwrap()
+                        .into_element();
+
+                    let create_stmt = match &mut stmt {
+                        Statement::CreateSink(s) => s,
+                        _ => coord_bail!("sink {id} was not created with a CREATE SINK statement"),
+                    };
+
+                    let new_config = match (&old_sink.host_config, size, remote) {
```
[the remark this answers is not in the document]
Comment (andrioni) [reply src/adapter/src/catalog.rs:3489]: I'm thinking about it! When #15312 gets merged, this logic could be unified with [this one](https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/plan/statement/ddl.rs#L1130-L1140), as we should be able to drop `RESET` and replace `AlterOptionParameter` (née `AlterSourceItem`) with just an `Option`.

[the remark this answers is not in the document]
Comment (andrioni) [reply src/adapter/src/catalog.rs:3489]: Gave it a try anyway!

**Comment by benesch** [comment]
@andrioni—do you think this can make the release tomorrow?

**Comment by andrioni** [comment]
PR is ready, now with proper k8s tests for custom sink sizes + ALTER SINK!

File: misc/python/materialize/cloudtest/wait.py [hunk 5]
```diff
@@ -9,13 +9,19 @@
...
 from materialize.ui import UIError
 
 
 def wait(
-    condition: str, resource: str, timeout_secs: int = 300, context: str = "kind-kind"
+    condition: str,
+    resource: str,
+    timeout_secs: int = 300,
+    context: str = "kind-kind",
+    *,
```
Comment (bkirwi) [inline misc/python/materialize/cloudtest/wait.py:23]: TIL!

File: test/cloudtest/test_storaged.py [hunk 6]
```diff
@@ -147,3 +147,71 @@ def test_storaged_shutdown(mz: MaterializeApplication) -> None:
...
+              SET (SIZE '16');
+            """
+        ),
+        no_reset=True,
+    )
+
+    wait(
+        condition="condition=Ready",
+        resource="pod",
+        label="storage.environmentd.materialize.cloud/size=16",
```
Comment (bkirwi) [inline test/cloudtest/test_storaged.py:211]: Oh interesting -- I was struggling to get this to work previously for another test!

**Review by bkirwi — APPROVED** [review]
Test and refactorings look good!

File: misc/python/materialize/cloudtest/wait.py (hunk 5 above)
Comment (andrioni) [reply misc/python/materialize/cloudtest/wait.py:23]: One of the lesser known features introduced with Python 3! I really like keyword-only args for cases like this.

**Pull request merged**